### PR TITLE
update block gossip validation when execution payload is delayed

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description
- Updates Gloas block gossip validation to distinguish between blocks building on an EMPTY parent and blocks building on a FULL parent execution payload
- Previously we required every `execution_payload_bid.parent_block_hash` to already be known by the execution layer, which could incorrectly reject valid Gloas blocks and downscore peers
- Now, EMPTY parent blocks are accepted when `parent_execution_requests` are empty, while FULL parent blocks wait for the missing parent execution payload with `SAVE_FOR_FUTURE`.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

---

> [!NOTE]
> **Medium Risk**
> Changes consensus-adjacent gossip validation for Gloas EPBS; incorrect EMPTY/FULL classification could accept bad blocks or reject/downscore peers, though behavior is aligned with block processing and heavily tested.
> 
> **Overview**
> **Gloas block gossip** no longer treats every `execution_payload_bid.parent_block_hash` as something the execution layer must already know. Validation now mirrors `process_parent_execution_payload`: **FULL** parents (bid hash matches `latest_execution_payload_bid`) require matching `parent_execution_requests` and return **`SAVE_FOR_FUTURE`** until the parent payload hash is known; **EMPTY** parents (hash matches `latest_block_hash` but not the latest bid) accept blocks only when `parent_execution_requests` are empty. FULL is checked before EMPTY so fork-transition cases where both hashes align are handled correctly.
> 
> Shared predicates live in **`MiscHelpersGloas`** (`isExecutionPayloadBidForFullParent`, `isExecutionPayloadBidForEmptyParent`, `isEmptyExecutionRequests`, `isExecutionRequestsRootMatchingLatestExecutionPayloadBid`), used by **`BlockProcessorGloas`** and gossip. Tests cover fork entry, delayed full payloads, empty-parent acceptance/rejection, and overlapping hash edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 28d609bdb73394184fd6463f4a52ce6ed5bb03f9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java [hunk 1]
```diff
@@ -240,6 +243,32 @@ public boolean isActiveBuilder(final BeaconState state, final UInt64 builderInde
     return predicates.isActiveBuilder(state, builderIndex);
   }
 
+  public boolean isExecutionPayloadBidForEmptyParent(
```
Comment (zilm13) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java:246]: if I correctly understands it will be true both for full and empty. I see that usage is after checking that it's not full, but it's common helper which could be used later by other person. I'd extend the check or rename or change flow.

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java [hunk 2]
```diff
@@ -240,6 +243,32 @@ public boolean isActiveBuilder(final BeaconState state, final UInt64 builderInde
...
+    return bid.getParentBlockHash().equals(BeaconStateGloas.required(state).getLatestBlockHash());
+  }
+
+  public boolean isExecutionPayloadBidForFullParent(
+      final BeaconState state, final ExecutionPayloadBid bid) {
+    return bid.getParentBlockHash()
+        .equals(BeaconStateGloas.required(state).getLatestExecutionPayloadBid().getBlockHash());
+  }
+
+  public boolean isDefaultExecutionRequests(final ExecutionRequests executionRequests) {
```
Comment (zilm13) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java]: could we reuse it? I think I saw this check in other place(s)

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java [hunk 3]
```diff
@@ -240,6 +243,32 @@ public boolean isActiveBuilder(final BeaconState state, final UInt64 builderInde
...
+    return executionRequests.equals(
+        schemaDefinitionsGloas.getExecutionRequestsSchema().getDefault());
+  }
+
+  public boolean isExecutionRequestsRootMatchingLatestExecutionPayloadBid(
```
Comment (zilm13) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java:265]: And this check looks common too
I'd find places where it could be reused and wire up

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java [hunk 4]
```diff
@@ -251,20 +248,10 @@ private InternalValidationResult performStatefulValidation(
...
-          executionPayloadBid.getParentBlockHash(), block.getParentRoot())) {
-        if (isParentFullPayloadDependency(parentBlockSlot, parentState, executionPayloadBid)) {
-          LOG.trace(
-              "Block requires parent execution payload that is not available. Saving for future processing.");
-          return InternalValidationResult.SAVE_FOR_FUTURE;
-        }
-        return reject(
-            "The parent block hash %s from the bid is not present or hasn't been passed validation",
-            executionPayloadBid.getParentBlockHash());
+      final Optional<InternalValidationResult> maybeParentExecutionPayloadValidationResult =
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java:257]: Please, restore the comment of what we check here

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java [hunk 5]
```diff
@@ -296,18 +283,78 @@ private InternalValidationResult equivocationCheckResultToInternalValidationResu
...
+        .miscHelpers()
         .toVersionGloas()
-        .map(BeaconStateGloas::getLatestExecutionPayloadBid)
-        .map(parentBid -> parentBid.getBlockHash().equals(executionPayloadBid.getParentBlockHash()))
-        .orElse(false);
+        .map(
+            miscHelpersGloas ->
+                validateGloasExecutionPayloadBidParent(
+                    block, parentState, executionPayloadBid, miscHelpersGloas))
+        .orElseGet(() -> validateKnownParentBlockHash(executionPayloadBid, block.getParentRoot()));
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java]: I maybe wrong but i don't see any case when we could reach these lines
if it's only for first block in Gloas we should modify names to make it clearly visible and have an appropriate test case

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java (hunk 1 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java:246]: I made `EMPTY` exclusive so it now matches `latest_block_hash` only when it does not also match `latest_execution_payload_bid.block_hash`. Added a regression test for that too

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java (hunk 2 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java]: Done. `BlockProcessorGloas` now uses the helper

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java (hunk 3 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java:265]: Done. `BlockProcessorGloas` now reuses the helper

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java (hunk 4 above)
Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java:257]: Done

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java (hunk 5 above)
Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java]: removed the fallback and now explicitly require Gloas helpers from the block slot

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java [hunk 6]
```diff
@@ -296,18 +289,59 @@ private InternalValidationResult equivocationCheckResultToInternalValidationResu
...
-    if (spec.atSlot(parentBlockSlot).getMilestone().isLessThan(SpecMilestone.GLOAS)) {
-      return false;
+    return validateGloasExecutionPayloadBidParent(
+        block,
+        parentState,
+        executionPayloadBid,
+        MiscHelpersGloas.required(spec.atSlot(block.getSlot()).miscHelpers()));
+  }
+
+  private Optional<InternalValidationResult> validateGloasExecutionPayloadBidParent(
```
Comment (tbenr) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java]: i'd remove `gloas` from the method .. seems like we can move the whole thing into the validateExecutionPayloadBidParent

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java [hunk 7]
```diff
@@ -240,6 +243,35 @@ public boolean isActiveBuilder(final BeaconState state, final UInt64 builderInde
...
+            .equals(stateGloas.getLatestExecutionPayloadBid().getBlockHash());
+  }
+
+  public boolean isExecutionPayloadBidForFullParent(
+      final BeaconState state, final ExecutionPayloadBid bid) {
+    return bid.getParentBlockHash()
+        .equals(BeaconStateGloas.required(state).getLatestExecutionPayloadBid().getBlockHash());
+  }
+
+  public boolean isDefaultExecutionRequests(final ExecutionRequests executionRequests) {
```
Comment (tbenr) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java]: maybe we can logically expose this as "isEmptyExecutionRequests" (it is the default ssz object for that schema byt at application level this represents an empty, which matches the concept of emptiness in gloas)

**Review by tbenr — DISMISSED** [review]
LGTM

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java (hunk 6 above)
Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java]: done

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java (hunk 7 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/MiscHelpersGloas.java]: done

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java (hunk 6 above)
Comment (tbenr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java]: I ment we could remove the wrapper method above.

Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java]: My bad! Fixed now

**Review by tbenr — DISMISSED** [review]
LGTM

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java [hunk 8]
```diff
@@ -296,18 +289,49 @@ private InternalValidationResult equivocationCheckResultToInternalValidationResu
...
-        .toVersionGloas()
-        .map(BeaconStateGloas::getLatestExecutionPayloadBid)
-        .map(parentBid -> parentBid.getBlockHash().equals(executionPayloadBid.getParentBlockHash()))
-        .orElse(false);
+
+    final ExecutionRequests parentExecutionRequests = maybeParentExecutionRequests.get();
+    // Gloas process_parent_execution_payload treats a parent as FULL when the child bid references
+    // the latest committed FULL parent bid. Check this before EMPTY because the FULL bid hash can
+    // equal latest_block_hash at Gloas genesis or fork transition.
+    if (miscHelpersGloas.isExecutionPayloadBidForFullParent(parentState, executionPayloadBid)) {
```
[the remark this answers is not in the document]
Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java:308]: the block gossip validation retrieves the parent state via `getParentStateInBlockEpoch`, so when the child block is in the first Gloas epoch and the parent block is pre Gloas, the retrieved state is regenerated at the child epoch start and is then a Gloas state

**Review by tbenr — APPROVED** [review]
LGTM
